### ISSUE TITLE
Export documentation items

### ISSUE THREAD

**Issue by SteinHeselmans** [melexis/sphinx-traceability-extension#31]
The database of documentation items could be exported to file. This allows import into other tools.

As an exported format the ReqIf format could be used: https://en.wikipedia.org/wiki/Requirements_Interchange_Format

**Comment by SteinHeselmans** [comment]
a preliminary start is made in #49

**Comment by SteinHeselmans** [comment]
first release of experimental json export in V2.5.0

**Comment by JasperCraeghs** [comment]
Do we want to continue using JSON? 
What item information needs to be added? 
Why is the current output a list of dictionaries instead of one dictionary with item IDs as keys?

**Comment by SteinHeselmans** [comment]
json, i don't know. It's just a format that i picked at that time. I've set it up to allow for multiple export formats to be created. I think ReqIf is definitely one to look at, as that is understood by other requirements tools.

The content of an item is not added to the export, since it is RST-syntax and probably contains tables, or whatever plain RST syntax which is not readable by other importing tools... I couldn't think for a solution, that's why the export is still preliminary :wink: 

A list-of-dict's iso dict's? I have no clue. Your suggestion makes sense however. It should be easy to convert the exported format, and i believe nobody is using this export feature till now :+1:

**Comment by bavovanachte** [comment]
Does a templated approach make sense? We could include a set of default templates for generating a simple json or the reqif format, but also allow the user to specify their own template. In that case, a user could choose to include the body of the item (and convert it to html, latex,...) if they choose to.

**Comment by sannenj** [comment]
I like the idea of a requirements export to ReqIF, but I would probably not use it. 
However I would love to see a ReqIF import possibility...

**Comment by SteinHeselmans** [comment]
templates do make sense, make sure to check our https://github.com/melexis/json-to-mako which we could use if we'ld use the json as an intermediate step.

For pyreqif i found https://github.com/ebroecker/pyreqif a while ago but didn't do any experiments. So i'm not sure if a templated approach would (work at all, or) be better than using this dedicated library.

Importing also seems a valid feature we need to keep an eye on.

**Comment by bavovanachte** [comment]
As background information: there would be 2 main use cases for this in our project:

1. Allow exporting the requirements into other tools.
2. Allow diff'ing a set of requirements/items between 2 versions of the documentation. 

For the first one, ReqIf is an excellent candidate, but it likely makes diff'ing 2 versions tricky due to it being XML. The current json output might be better suited except for the fact that it doesn't allow detecting whether a requirement's content has changed between 2 versions. It might be interesting to include a content hash for that purpose.

**Comment by bavovanachte** [comment]
@SteinHeselmans did you do any further investigations into https://github.com/ebroecker/pyreqif? Looking at the repo, it doesn't look like a very well maintained repository.

**Comment by SteinHeselmans** [comment]
no i didn't go further than the readme which indeed doesn't promise too much (export is buggy), and >2y no commit is also a bad sign.

Altogether I think json is the better option for now

**Comment by bavovanachte** [comment]
TL;DR: Proposal to split this into 2 issues:
- Integrate reqif export/(import?)
- Add content hash to existing json output

An update on how we are currently using the json output:

The json file is used to generate the contents for FMEA analysis: using a python script we "reconstruct" the rqt tree in order to identify the top-level requirements assigned to a specific SW component (these can be identified using the rqt ID). These are then exported along with some FMEA-specific content as a csv which can be imported in to the FMEA spreadsheet infrastructure.

This all works well if it's done once. When iterating over the FMEA process, we would need to be able to merge a new export with the existing work in the FMEA sheet. New/removed requirements can be easily identified by looking at the list of the exported requirements. 

What is not possible at the moment is identifying requirements that have **changed** between 2 versions of the export. A proposal would be to include a hash of the content in the current json output. This can then be used to see which requirements have changed contents.

I believe the ReqIf export still makes sense for a number of use cases (exchanging requirements with existing tools). I would propose to split this issue up into 2 separate ones, content hash on shorter term and ReqIf on longer term.

Opinions? @SteinHeselmans @JasperCraeghs

**Comment by JasperCraeghs** [comment]
Your proposal seems good to me. I won't have time to work on this for about another two weeks, however. Please feel free to contribute.

**Comment by bavovanachte** [comment]
The content hash was added in #158 . I'll close this ticket and open a new one for the ReqIf export

**Comment by bavovanachte** [comment]
Opened #164 for ReqIf export
